src/lib.rs

```diff
@@ -0,0 +1,545 @@
+fn check_that_maps_match(
+    old_map: &BTreeMap<i64, StateGroupEntry>,
+    new_map: &BTreeMap<i64, StateGroupEntry>,
+) {
+    println!("Checking that state maps match...");
+
+    let pb = ProgressBar::new(old_map.len() as u64);
+    pb.set_style(
+        ProgressStyle::default_bar().template("[{elapsed_precise}] {bar} {pos}/{len} {msg}"),
+    );
+    pb.set_message("state groups");
+    pb.enable_steady_tick(100);
+
+    // Now let's iterate through and assert that the state for each group
+    // matches between the two versions.
+    old_map
+        .par_iter() // This uses rayon to run the checks in parallel
+        .try_for_each(|(sg, _)| {
+            let expected = collapse_state_maps(&old_map, *sg);
+            let actual = collapse_state_maps(&new_map, *sg);
+
+            pb.inc(1);
+
+            if expected != actual {
+                println!("State Group: {}", sg);
+                println!("Expected: {:#?}", expected);
+                println!("actual: {:#?}", actual);
+                Err(format!("State for group {} do not match", sg))
+            } else {
+                Ok(())
+            }
+        })
+        .expect("expected state to match");
+
+    pb.finish();
+
+    println!("New state map matches old one");
+}
+
+/// Gets the full state for a given group from the map (of deltas)
+fn collapse_state_maps(map: &BTreeMap<i64, StateGroupEntry>, state_group: i64) -> StateMap<Atom> {
+    let mut entry = &map[&state_group];
+    let mut state_map = StateMap::new();
+
+    let mut stack = vec![state_group];
+
+    while let Some(prev_state_group) = entry.prev_state_group {
+        stack.push(prev_state_group);
+        if !map.contains_key(&prev_state_group) {
+            panic!("Missing {}", prev_state_group);
+        }
+        entry = &map[&prev_state_group];
+    }
+
+    for sg in stack.iter().rev() {
+        state_map.extend(
+            map[&sg]
+                .state_map
+                .iter()
+                .map(|((t, s), e)| ((t, s), e.clone())),
+        );
+    }
+
+    state_map
+}
+
+// PyO3 INTERFACE STARTS HERE
+
+impl Config {
+    /// Converts string and bool arguments into a Config struct
+    ///
+    /// This function panics if db_url or room_id are empty strings!
+    pub fn new(
+        db_url: String,
+        output_file: String,
+        room_id: String,
+        max_state_group: String,
+        min_saved_rows: String,
+        transactions: bool,
+        level_sizes: String,
+    ) -> Config {
+        if db_url.is_empty() {
+            panic!("db url is required");
+        }
+
+        let mut output: Option<File> = None;
+        if !output_file.is_empty() {
+            output = Some(
+                File::create(output_file)
+                    .unwrap_or_else(|e| panic!("Unable to create output file: {}", e)),
+            );
+        }
+        let output_file = output;
+
+        if room_id.is_empty() {
+            panic!("room_id is required");
+        }
```

Comment:
I think we should drop this check.

Reply:
Have done, however the compressor is guaranteed to fail if either the room_id or db_url aren't set (is it not better to fail earlier and give a more descriptive error message?)

Reply:
> is it not better to spot this early and return an Err Result instead of dropping the check?

Ish. What concerns me in this particular case is that it hints that the empty string is being used to mean "not set", rather than checking if the room_id/db_url are valid. It also doesn't really help the caller much, by checking that those params are non-null isn't enough to be able to say e.g. "if Config::new works then the params are valid" or anything like that, so at that point you may as well leave it to the compressor to error if they're wrong (and if they're not valid the compressor will error quickly). The alternative is to actually parse the room ID/db connection param here, or maybe even connect to the DB to ensure it works.

Reply:
In a lot of ways, its *less* helper to API users if there are multiple different places we validate and may error if incorrect params are given.